students/srepking/Lesson09/donors.py

```diff
@@ -0,0 +1,140 @@
+                tot_wi = (len(str(summary[0]))) + 3
+                # width of number of donations column
+        return tot_wi
+
+    @staticmethod
+    def column_average_width(donor_summary):
+        ave_wi = 12
+        for name, summary in donor_summary.items():
+            if len(str(summary[2])) > ave_wi:
+                # width of total column
+                ave_wi = (len(str(summary[2]))) + 3
+                # width of number of donations column
+        return ave_wi
+
+    @staticmethod
+    def column_number_width(donor_summary):
+        num_wi = 12
+        for name, summary in donor_summary.items():
+            if len(str(summary[1])) > num_wi:
+                # width of total column
+                num_wi = (len(str(summary[1]))) + 3
+                # width of number of donations column
+        return num_wi
+
+    @staticmethod
+    def sort_list(donor_summary):
+        list_sorted = sorted(donor_summary, key=donor_summary.__getitem__, reverse=True)
+        return list_sorted
+
+    def make_report(self):
+        """Return a report on all the donors"""
+        donor_summary = self.summary()
+        name_wi = Group.column_name_width(donor_summary)
+        tot_wi = Group.column_total_width(donor_summary)
+        num_wi = Group.column_number_width(donor_summary)
+        ave_wi = Group.column_average_width(donor_summary)
+
+        list_sorted = Group.sort_list(donor_summary)
+
+        rows = ['\n''A summary of your donors donations:',
+                f"{'Donor Name':{name_wi}}| {'Total Given':^{tot_wi}}| "
+                f"{'Num Gifts':^{num_wi}}| {'Average Gift':^{ave_wi}}",
+                f"{'-':-^{(name_wi+tot_wi+ave_wi+num_wi+8)}}"]
+
+        for key in list_sorted:
+            temp = donor_summary[key]
+            rows.append(f"{key:{name_wi}}${temp[0]:{tot_wi}.2f}"
+                        f"{temp[1]:^{num_wi}}   "
+                        f"${temp[2]:>{ave_wi}.2f}")
+
+        return '\n'.join(rows)
+
+    report = property(make_report)
+
+    def letters(self):
+        donors_f = self.summary()
+
+        for donor, donation in self._donor_raw.items():
+            donation_summary = donors_f[donor]
+            letter = f'Dear {donor}, thank you so much for your ' \
+                     f'last contribution of ${donation[-1]:.2f}! ' \
+                     f'You have contributed a total of $' \
+                     f'{donation_summary[0]:.2f}, ' \
+                     f'and we appreciate your support!'
+            # Write the letter to a destination
+            with open(donor + '.txt', 'w') as to_file:
+                to_file.write(letter)
+
+
+class Individual:
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def thank_you(donor, donation):
+        """Add a donation to a donors records and print a report."""
+        return ('Thank you so much for the generous gift of ${0:.2f}, {1}!'
+              .format(float(donation), donor))
```

Comment:
This class should be utilized in your `Group` class.

The relation ship here is one to many; you have `Donor` and `DonorCollection` - ideally `DonorCollection` would reference many `Donor` objects.

Here is an example of what I mean:
https://repl.it/@naaleksandrova/Lesson9Example

Reply:
In your example, you also have the thank you letter in the Donor Class. Let me know if this comment still stands after my next pull request. I have incorporated your example, and now my DonorCollection is referencing many Donor objects. This is a much nicer way of doing this! Thank you!